Add support for deployment to GKE

henrybell opened the pull request:
This PR restructures the repository and adds code/config to support deployment of the Autoscaler to Kubernetes (GKE) as well as to Google Cloud Functions, fixing #67. As well as this, #71 and #62 are additionally fixed as a result of changes to the service account provisioning/permissions within the Terraform code.

Because this PR is large and touches a significant proportion of the codebase, thorough testing before merging is recommended -- we can set aside some time for this if helpful.

bgood commented:
Thanks for the PR!

I've started working through it, but you are right it is a big one.  One thing we'll need to update are the GitHub Actions to run unit tests. I think the change is simply to update the working directories.

henrybell commented:
I fixed the `markdownlinter` findings and modified the paths for the unit tests -- will check the results next time the workflows run (currently waiting for approval). Thanks!

bgood commented:
Did a quick run through of the per-project Cloud Functions deployment.  Looks like the working directory exports in the READMEs need to be updated.  Here is an example from the per-project README, `export AUTOSCALER_DIR="$(pwd)/autoscaler/terraform/per-project"`.

- [ ] terraform/cloud-functions/distributed/README.md
- [ ] terraform/cloud-functions/per-project/README.md

bgood commented on README.md:
```diff
@@ -213,23 +137,24 @@ covered by the Google Cloud Spanner product support.
-[spanner-max-storage]: https://cloud.google.com/spanner/docs/monitoring-cloud#storage
-[cloud-scheduler]: https://cloud.google.com/scheduler
-[cloud-pub-sub]: https://cloud.google.com/pubsub
+[autoscaler-poller]: poller/README.md
```
Unused links.

bgood commented on terraform/gke/README.md:
```diff
@@ -0,0 +1,450 @@
+    ·
+    Terraform configuration
+    ·
+    Monitoring
```
Add link back to monitoring section.

bgood reviewed (changes requested):
Looking good. Made a few suggestions around cleaning links in the markdown.

henrybell commented on terraform/gke/README.md:
````diff
@@ -0,0 +1,440 @@
+    database, set the following variable:
+
+    ```sh
+    export TF_VAR_terraform_spanner_state=true
````
I have these docs changes made on another branch, will add to the PR once tested.

bgood commented:
@davidcueva , this change is looking good.  Given the size and scope would you also be able to do a quick review?

henrybell commented:
Closing to recreate from new `master` branch following #75.
